### PR TITLE
[PM-3758] Handle user decryption options from pre-TDE server response

### PR DESCRIPTION
**Pull request by trmartin4**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Prior to TDE, the user's "decryption options" were implied based on the presence or absence of the `keyConnectorUrl` in the token response.  Post-TDE, we have build an explicit `UserDecryptionOptions` property on the response from the server.  However, clients must be able to handle pre-TDE server versions for our self-hosted customers who do not upgrade. 

This has been done by using the `keyConnectorUrl` to initialize the `hasMasterPassword`.  The key code to look at on the server is [here](https://github.com/bitwarden/server/blob/5f4a303180ea4f6caf9c5ed50c6c3858e678e18b/src/Identity/IdentityServer/CustomTokenRequestValidator.cs#L104), in the pre-TDE server-side code.  If the user has a master password, the `keyConnectorUrl` will **not** be in the response.  This allows us to infer the `hasMasterPassword` property based on the `keyConnectorUrl` being present.

## Code changes

- **account.ts:** Modified the `fromJson` for the `AccountDecryptionOptions` to handle mapping the `hasMasterPassword` and `keyConnectorUrl` from the pre-TDE server response.  This required changing the input to include the entire `IdentityTokenResponse` object.
- **login.strategy.ts:** Modified the `fromJson` to pass in the full `IdentityTokenResponse` object.
- **sso-login.strategy.ts:** Simplified the logic in `shouldSetMasterKeyFromKeyConnector` to be based on the presence of `UserDecryptionOptions` in the response.  This parallels the code flow in `account.ts` and keeps it consistent and easier to remove in the future when we can rip out the old response structure.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/common/src/auth/login-strategies/sso-login.strategy.spec.ts [hunk 1]
```diff
@@ -266,7 +266,61 @@ describe("SsoLogInStrategy", () => {
...
     beforeEach(() => {
-      tokenResponse = identityTokenResponseFactory(null, { HasMasterPassword: false });
+      tokenResponse = identityTokenResponseFactory(null, {
+        HasMasterPassword: false,
+        KeyConnectorOption: { KeyConnectorUrl: keyConnectorUrl },
+      });
+      tokenResponse.keyConnectorUrl = keyConnectorUrl;
+    });
+
+    it("gets and sets the master key if Key Connector is enabled and the user doesn't have a master password", async () => {
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/login-strategies/sso-login.strategy.spec.ts:276]: Minor nit (non-blocking): we have new test name conventions defined here https://contributing.bitwarden.com/contributing/testing/unit/naming-conventions

File: libs/common/src/auth/login-strategies/sso-login.strategy.spec.ts
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/login-strategies/sso-login.strategy.spec.ts:1]: Thank you for adding tests for these!

File: libs/common/src/platform/models/domain/account.ts [hunk 2]
```diff
@@ -311,28 +311,46 @@ export class AccountDecryptionOptions {
   //   return this.keyConnectorOption !== null && this.keyConnectorOption !== undefined;
   // }
 
-  static fromResponse(response: UserDecryptionOptionsResponse): AccountDecryptionOptions {
+  static fromResponse(response: IdentityTokenResponse): AccountDecryptionOptions {
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/platform/models/domain/account.ts:314]: Good call on fixing this here so we can still use the `accountDecryptionOptions` elsewhere. Thank you for fixing this missed scenario!

**Review by JaredSnider-Bitwarden — APPROVED** [review]
LGTM! Thank you for fixing this. 1 minor non blocking nit pick.

File: libs/common/src/platform/models/domain/account.ts (hunk 2 above)
Comment (justindbaur) [inline libs/common/src/platform/models/domain/account.ts:314]: In the future we should work on either moving the whole class `AccountDecryptionOptions` into auth owned code or at least have the code in `fromResponse` into a function that we import here from auth code. The logic in there doesn't feel to me like platform should be owning it.

**Review by justindbaur — APPROVED** [review]
Everything looks great, especially the new tests!

File: libs/common/src/platform/models/domain/account.ts (hunk 2 above)
Comment (trmartin4) [reply libs/common/src/platform/models/domain/account.ts:314]: Created https://bitwarden.atlassian.net/browse/PM-3778.

**Pull request merged**